example/example.d

```diff
@@ -88,4 +111,64 @@ void main(string[] args)
     // Signal that the COPY is finished. Let Postgresql finalize the command
     // and return any errors with the data.
     conn.putCopyEnd();
+    writeln("CSV copy-in worked.");
+
+
+    // It is also possible to send raw binary data.  It's even faster, and can
+    // handle any PostgreSQL type, including BYTEA, but it's more complex than
+    // sending parsable text streams
+    conn.exec("CREATE TEMP TABLE test_dpq2_blob (item BIGINT, data BYTEA)");
+
+    // Init the COPY command, this time for direct binary input
+    conn.exec("COPY test_dpq2_blob (item, data) FROM STDIN WITH (FORMAT binary)");
+
+    // For FORMAT binary, send over the 19 byte PostgreSQL header manually
+    //                   P    G    C    O    P    Y   \n  255   \r   \n
+    conn.putCopyData(cast(ubyte[])[
+        0x50,0x47,0x43,0x4F,0x50,0x59,0x0A,0xFF,0x0D,0x0A,0,0,0,0,0,0,0,0,0
+    ]);
+
+    // Write 10 rows of variable length binary data. PostgreSQL internal 
+    // storage is big endian and the number of values and the length of each
+    // must be provided.  Since binary copy-in is likely to be used in 
+    // "tight-loop" code, we'll use a stack memory. Stack buffer size is:
+    // 
+    //   2 bytes for the number of values
+    //   plus 4 length bytes for each value
+    //   plus total size of all values in the largest row
+    //
+    enum LOOPS = 10;
+    ubyte[2 + 2*4 + 8 + (2*LOOPS + 7)] buf;
+    foreach(i;0..LOOPS){
+        size_t offset = 0;
+        buf[].write!(short, BE)(2, &offset);            // Sending two fields
+
+        buf[].write!(int,   BE)(long.sizeof, &offset);  // BIGINT == long
```

Comment:
In the `dpq2.conv.from_d_types` module contains converter for (almost) all supported types. Really, it puts D values into `Value` struct, but it is possible to extract needed code into a separate codeblock